File: abridgedreading.py
from sys import stdin,stdout

n,m = map(int,stdin.readline().split())
pages = []
while len(pages) < n:
    lst = list(map(int,stdin.readline().split()))
    pages.extend(lst)
adjList,CC = [-1 for _ in range(n)],{i for i in range(n)}
for i in range(m):
    u,v = map(lambda x: int(x) - 1, stdin.readline().split())
    #every vertex has at most one parent
    adjList[v] = u #the edges directions are reversed
    CC.discard(u)
#CC contains list of possible culminating concept chapters
CC,lowest = list(CC),10**9
#O(n^2) worst case occurs when all chapters are culminating concept chapters
for i in range(len(CC)):
    u = CC[i]
    t,read = 0,set()
    while u != -1:
        read.add(u)
        t += pages[u]
        u = adjList[u]
    for j in range(i + 1,len(CC)):
        v,temp = CC[j],t
        while v != -1 and v not in read:
            temp += pages[v]
            v = adjList[v]   
        lowest = min(lowest,temp)
stdout.write(f"{lowest}\n")

File: abridgedreading.cpp
#include <bits/stdc++.h>
#include <unordered_set>

using namespace std;

typedef vector<int> vi;
typedef unordered_set<int> usi;

int main() {
    int n,m; scanf("%d %d",&n,&m);
    vi pages(n,0), pred(n,-1), outdeg(n,0), CC;
    for (int i = 0; i < n; i++) scanf("%d",&pages[i]);
    for (int i = 0; i < m; i++) {
        int a,b; scanf("%d %d",&a,&b);
        a--; b--;
        pred[b] = a; outdeg[a]++;
    }
    for (int i = 0; i < n; i++) {
        if (outdeg[i] == 0) CC.push_back(i);
    }
    int lowest = 1000000000;
    for (int i = 0; i < CC.size(); i++) {
        int t = 0; usi read;
        int u = CC[i];
        while (u != -1) {
            t += pages[u];
            read.insert(u);
            u = pred[u];
        }
        for (int j = i + 1; j < CC.size(); j++) {
            int v = CC[j], temp = t;
            while (v != -1 && read.find(v) == read.end()) {
                temp += pages[v];
                v = pred[v];
            }
            lowest = min(lowest,temp);
        }
    }
    printf("%d\n",lowest);
    return 0;
}